lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
+    Path(Path), OpenEC(), OS(Path, OpenEC), OutputtingArguments(false) {
+    revng_assert(!OpenEC);
+    printHeader();
+  }
+
+  ~PipelineCTracer() {
+    OS.flush();
+    OS.close();
+  }
+
+private:
+  void printHeader() {
+    OS << "version: 1\n";
+    OS << "commands:\n";
+    OS.flush();
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
+  }
+
+public:
+  void functionPrelude(const llvm::StringRef Name) {
+    OS << "- name: " << Name << "\n";
+    OS << "  arguments:\n";
+    OutputtingArguments = true;
+    OS.flush();
+  }
+
+  void newArgument() {
+    OS << "  - ";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t> ||
+           std::is_same_v<T, uint32_t> ||
+           std::is_same_v<T, uint64_t>
+  // clang-format on
+  void printValue(const T &Int) {
+    OS << Int << "\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, char *> ||
+           std::is_same_v<T, const char *>
+  // clang-format on
+  void printValue(T String) {
+    if (OutputtingArguments)
+      OS << reprString(String);
+    else
+      OS << "P" << (void *) String;
```

Comment:
1. No C-style casts
2. Cast to `intptr_t`, no `void *` anywhere
3. I'd prefer to have something like `object_1`, `object_2`, much easier to read. Can be done easily with a `std::map<intptr_t, unsigned>`.

Reply:
1. Removed all of them
2. Done
3. I'm really against this approach. Some considerations:
   1. Pointers make some things pretty obvious, e.g. P0x0 is NULL, P0x7f... is on the stack and P0x55.. is on the heap etc.
   2. The trace should be a reflection of what happened, what happened is that pointers were passed around
   3. `TraceWriter` should really do the minimal required work to make `trace run` functional. What if python conjures a pointer out of thin air as an argument? The `std::map` could cause an abort and then we'd miss it in the trace or we catch it in time and print a pointer, now there's 2 ways of representing pointer in the trace
   4. I have doubts the trace will be read many times, the most common thing will be a `run trace` with gdb followed by a whole lot of `print Foo.dump()`.
   5. The `object_` approach still suffers from aliasing, if I `_destroy` a pointer then it can be re-used, so you would see `object_20` halfway through the trace, which is confusing. If we handle aliasing we're back at point (3) with extra unneeded logic
   6. Any non-trivial trace will easily reach 10k objects, the legibility of `object_12345` isn't that great 
   7. If we still want to read traces, I'd advocate for a pretty-printer, e.g. `trace inspect --pretty-print` where we can implement as many "nice things" to visualize the trace as we wish without having to put in the tracing logic

Reply:
1. Not sure why we should care about heap vs stack. Fair point about `nullptr`, but still... If you really think that's important you can initialize the map with `{ nullptr, 0 }`.
2. Not sure why this should be an argument. If anything, `P0x1000` in `gdb` will have a different value, which makes things more confusing than helpful.
3. Well, if absent we can simply emit "object_invalid" and immediately stop (it's a bug, we should stop). This seems more of an argument *for* using `object_...`.
4. Uhm, not really an argument.
5. Pointers suffer from aliasing too. If anything with `object_xxx` we have the option of detecting functions that return new objects and functions that destroy objects and handle them properly avoiding aliasing. Not saying we have to do this now, but it leaves a door open.
6. Searching for pointers is for sure more difficult, either you copy/past or you have to type the last few digits, and you never know how many you need. Remembering 73512 is much easier than `0x00007ffff7fe3f30`.
7. You can always get more tooling/post-processing, but that's not an argument for making good things by default.

The bottom line is that pointers are not relevant in the trace, since they are not reproducible and therefore misleading while debugging (you have to remember to ignore them). Also, harder to read. The extra logic seems very very thin to me.

Anyway, this is not a matter of life or death. If you really really want to keep them this way at least use `ptr_` instead of `P` as a prefix.

Reply:
Changed prefix to `ptr_` and left note on the `object_` approach